File: code/packages/fsharp/vhdl-parser/VhdlParser.fs
namespace CodingAdventures.VhdlParser.FSharp

open System
open System.Collections.Concurrent
open System.IO
open System.Text
open CodingAdventures.GrammarTools.FSharp
open CodingAdventures.Parser.FSharp
open CodingAdventures.VhdlLexer.FSharp

/// VHDL parser backed by shared grammar definitions.
type VhdlParser private () =
    static let parserGrammars = ConcurrentDictionary<string, ParserGrammar>()

    static member DefaultVersion = VhdlLexer.DefaultVersion
    static member SupportedVersions = VhdlLexer.SupportedVersions

    static member CreateVhdlParser() = VhdlParser.CreateVhdlParser(VhdlParser.DefaultVersion)
    static member CreateVhdlParser(version: string) = GrammarParser(VhdlParser.LoadParserGrammar(version))

    static member ParseVhdl(source: string) = VhdlParser.ParseVhdl(source, VhdlParser.DefaultVersion)
    static member ParseVhdl(source: string, version: string) =
        try
            VhdlParser.CreateVhdlParser(version).Parse(VhdlLexer.TokenizeVhdl(source, version))
        with
        | :? GrammarParseError as error ->
            raise (ArgumentException(sprintf "VHDL parse failed: %s" error.Message, "source", error))

    static member private LoadParserGrammar(version: string) =
        let validated = VhdlParser.ValidateVersion(version)
        parserGrammars.GetOrAdd(validated, Func<string, ParserGrammar>(VhdlParser.ParseParserGrammarResource))

    static member private ValidateVersion(version: string) =
        if String.IsNullOrWhiteSpace(version) then
            VhdlParser.DefaultVersion
        elif VhdlParser.SupportedVersions |> Seq.exists ((=) version) then
            version
        else
            invalidArg "version" (sprintf "Unknown VHDL version '%s'. Valid values: %s" version (String.Join(", ", VhdlParser.SupportedVersions)))

    static member private ParseParserGrammarResource(version: string) =
        try
            ParserGrammarParser.Parse(VhdlParser.ReadResource("vhdl" + version + ".grammar"))
        with
        | :? ParserGrammarError as error ->
            raise (InvalidOperationException(sprintf "Failed to parse bundled VHDL parser grammar for version %s" version, error))

    static member private ReadResource(resourceName: string) =
        let assembly = typeof<VhdlParser>.Assembly
        use stream = assembly.GetManifestResourceStream(resourceName)
        if isNull stream then
            invalidOp (sprintf "Missing bundled resource: %s" resourceName)
        use reader = new StreamReader(stream, Encoding.UTF8)
        reader.ReadToEnd()

File: code/packages/csharp/vhdl-parser/VhdlParser.cs
using System.Collections.Concurrent;
using System.Text;
using CodingAdventures.GrammarTools;
using CodingAdventures.Parser;

namespace CodingAdventures.VhdlParser;

/// <summary>
/// VHDL parser backed by shared grammar definitions.
/// </summary>
public static class VhdlParser
{
    public static string DefaultVersion => CodingAdventures.VhdlLexer.VhdlLexer.DefaultVersion;
    public static IReadOnlyList<string> SupportedVersions => CodingAdventures.VhdlLexer.VhdlLexer.SupportedVersions;

    private static readonly ConcurrentDictionary<string, ParserGrammar> ParserGrammars = new();

    public static GrammarParser CreateVhdlParser() => CreateVhdlParser(DefaultVersion);

    public static GrammarParser CreateVhdlParser(string? version) => new(LoadParserGrammar(version));

    public static ASTNode ParseVhdl(string source) => ParseVhdl(source, DefaultVersion);

    public static ASTNode ParseVhdl(string source, string? version)
    {
        try
        {
            return CreateVhdlParser(version).Parse(CodingAdventures.VhdlLexer.VhdlLexer.TokenizeVhdl(source, version));
        }
        catch (GrammarParseError error)
        {
            throw new ArgumentException("VHDL parse failed: " + error.Message, nameof(source), error);
        }
    }

    private static ParserGrammar LoadParserGrammar(string? version)
    {
        var validated = ValidateVersion(version);
        return ParserGrammars.GetOrAdd(validated, ParseParserGrammarResource);
    }

    private static string ValidateVersion(string? version)
    {
        if (string.IsNullOrWhiteSpace(version))
        {
            return DefaultVersion;
        }

        if (!SupportedVersions.Contains(version))
        {
            throw new ArgumentException("Unknown VHDL version '" + version + "'. Valid values: " + string.Join(", ", SupportedVersions), nameof(version));
        }

        return version;
    }

    private static ParserGrammar ParseParserGrammarResource(string version)
    {
        try
        {
            return ParserGrammarParser.Parse(ReadResource("vhdl" + version + ".grammar"));
        }
        catch (ParserGrammarError error)
        {
            throw new InvalidOperationException("Failed to parse bundled VHDL parser grammar for version " + version, error);
        }
    }

    private static string ReadResource(string resourceName)
    {
        var assembly = typeof(VhdlParser).Assembly;
        using var stream = assembly.GetManifestResourceStream(resourceName)
            ?? throw new InvalidOperationException("Missing bundled resource: " + resourceName);
        using var reader = new StreamReader(stream, Encoding.UTF8);
        return reader.ReadToEnd();
    }
}
